// tdesign-mobile-vue/es/input/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import '../shared/index.js';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, ref, toRefs, computed, nextTick, watch, createVNode } from 'vue';
import { CloseCircleFilledIcon } from 'tdesign-icons-vue-next';
import { useFocus } from '@vueuse/core';
import config from '../config.js';
import InputProps from './props.js';
import { useFormDisabled } from '../form/hooks.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { useDefault } from '../shared/useDefault/index.js';
import { getCharacterLength } from '../shared/util.js';
import { extendAPI, withInstall } from '../shared/component.js';
import './style/css.js';
import './type.js';
import '../shared/functions.js';
import '../shared/constants.js';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import 'lodash/camelCase';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '../shared/useCountDown/utils.js';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useVModel/index.js';
import 'lodash/kebabCase';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import 'lodash/isArray';
import '../shared/useGesture/index.js';
import '@use-gesture/vanilla';
import '../shared/hover.js';
import '../config-provider/useConfig.js';
import 'lodash/isFunction';
import 'lodash/cloneDeep';
import 'lodash/isString';
import '../config-provider/context.js';
import 'lodash/mergeWith';
import 'lodash/merge';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-4c6045f6.js';
import '../_chunks/dep-1d280204.js';
import 'dayjs';
import '../_chunks/dep-5101c61a.js';
import '../config-provider/type.js';
import '../hooks/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isObject';
import 'lodash/isNumber';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var name = "".concat(prefix, "-input");
var Input = defineComponent({
  name: name,
  props: _objectSpread(_objectSpread({}, InputProps), {}, {
    labelAlign: {
      type: String,
      default: "top"
    },
    pattern: {
      type: String
    },
    inputmode: {
      type: String,
      validator: function validator(val) {
        if (!val) return true;
        return ["search", "text", "none", "url", "tel", "email", "numeric", "decimal"].includes(val);
      }
    }
  }),
  setup: function setup(props, context) {
    var readerTNodeJSX = useTNodeJSX();
    var inputClass = usePrefixClass("input");
    var isDisabled = useFormDisabled();
    var inputRef = ref();
    var _toRefs = toRefs(props),
      autofocus = _toRefs.autofocus;
    var _useDefault = useDefault(props, context.emit, "value", "change"),
      _useDefault2 = _slicedToArray(_useDefault, 1),
      innerValue = _useDefault2[0];
    var status = props.status || "default";
    var _useFocus = useFocus(inputRef, {
        initialValue: props.autofocus
      }),
      focused = _useFocus.focused;
    var inputClasses = computed(function () {
      return ["".concat(inputClass.value, "__control"), _defineProperty(_defineProperty(_defineProperty({}, "".concat(inputClass.value, "--").concat(props.align), props.align !== "left"), "".concat(inputClass.value, "--").concat(status), status), "".concat(inputClass.value, "__control--disabled"), isDisabled.value)];
    });
    var rootClasses = computed(function () {
      return [inputClass.value, _defineProperty(_defineProperty(_defineProperty({}, "".concat(inputClass.value, "--layout-").concat(props.layout), props.layout), "".concat(inputClass.value, "--center"), props.labelAlign === "center"), "".concat(inputClass.value, "--border"), !props.borderless)];
    });
    var setInputValue = function setInputValue() {
      var v = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var input = inputRef.value;
      var sV = String(v);
      if (!input) {
        return;
      }
      if (input.value !== sV) {
        input.value = sV;
      }
    };
    var handleInput = function handleInput(e) {
      if (e instanceof InputEvent) {
        var checkInputType = e.inputType && e.inputType === "insertCompositionText";
        if (e.isComposing || checkInputType) return;
      }
      inputValueChangeHandle(e);
    };
    var inputValueChangeHandle = function inputValueChangeHandle(e) {
      var value = e.target.value;
      var allowInputOverMax = props.allowInputOverMax,
        maxcharacter = props.maxcharacter;
      if (!allowInputOverMax && maxcharacter && maxcharacter > 0 && !Number.isNaN(maxcharacter)) {
        var _getCharacterLength = getCharacterLength(value, maxcharacter),
          _getCharacterLength$l = _getCharacterLength.length,
          length = _getCharacterLength$l === void 0 ? 0 : _getCharacterLength$l,
          _getCharacterLength$c = _getCharacterLength.characters,
          characters = _getCharacterLength$c === void 0 ? "" : _getCharacterLength$c;
        innerValue.value = characters;
      } else {
        innerValue.value = value;
      }
      nextTick(function () {
        return setInputValue(innerValue.value);
      });
    };
    var focus = function focus() {
      focused.value = true;
    };
    var blur = function blur() {
      focused.value = false;
    };
    extendAPI({
      focus: focus,
      blur: blur
    });
    var handleClear = function handleClear(e) {
      var _props$onClear;
      innerValue.value = "";
      focused.value = true;
      (_props$onClear = props.onClear) === null || _props$onClear === void 0 || _props$onClear.call(props, {
        e: e
      });
    };
    var handleFocus = function handleFocus(e) {
      var _props$onFocus;
      (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, innerValue.value, {
        e: e
      });
    };
    var handleBlur = function handleBlur(e) {
      var _props$onBlur;
      (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, innerValue.value, {
        e: e
      });
    };
    var handleCompositionend = function handleCompositionend(e) {
      inputValueChangeHandle(e);
    };
    watch(autofocus, function (autofocus2, prevAutofocus) {
      if (autofocus2 === true) {
        nextTick(function () {
          focused.value = true;
        });
      }
    });
    return function () {
      var readerPrefix = function readerPrefix() {
        var prefixIcon = readerTNodeJSX("prefixIcon");
        var label = readerTNodeJSX("label");
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--prefix")
        }, [prefixIcon && createVNode("div", {
          "class": "".concat(inputClass.value, "__icon--prefix")
        }, [prefixIcon]), createVNode("div", {
          "class": "".concat(inputClass.value, "__label")
        }, [label])]);
      };
      var readerClearable = function readerClearable() {
        if (props.clearable && innerValue.value && innerValue.value.length > 0) {
          return createVNode("div", {
            "class": "".concat(inputClass.value, "__wrap--clearable-icon"),
            "onClick": handleClear
          }, [createVNode(CloseCircleFilledIcon, null, null)]);
        }
        return null;
      };
      var readerSuffix = function readerSuffix() {
        var suffix = readerTNodeJSX("suffix");
        if (!suffix) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--suffix")
        }, [suffix]);
      };
      var readerSuffixIcon = function readerSuffixIcon() {
        var suffixIcon = readerTNodeJSX("suffixIcon");
        if (!suffixIcon) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--suffix-icon")
        }, [suffixIcon]);
      };
      var readerTips = function readerTips() {
        var tips = readerTNodeJSX("tips");
        if (!tips) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(inputClass.value, "__tips ").concat(inputClass.value, "--").concat(props.align)
        }, [tips]);
      };
      return createVNode("div", {
        "class": rootClasses.value
      }, [readerPrefix(), createVNode("div", {
        "class": "".concat(inputClass.value, "__wrap")
      }, [createVNode("div", {
        "class": "".concat(inputClass.value, "__content ").concat(inputClass.value, "--").concat(status || "default")
      }, [createVNode("input", {
        "ref": inputRef,
        "value": innerValue.value,
        "name": props.name,
        "class": inputClasses.value,
        "type": props.type,
        "disabled": isDisabled.value,
        "autocomplete": props.autocomplete ? "On" : "Off",
        "placeholder": props.placeholder,
        "readonly": props.readonly,
        "maxlength": props.maxlength || -1,
        "pattern": props.pattern,
        "inputmode": props.inputmode,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onCompositionend": handleCompositionend
      }, null), readerClearable(), readerSuffix(), readerSuffixIcon()]), readerTips()])]);
    };
  }
});

var _Input = withInstall(Input);

export { _Input as default };
