// tdesign-mobile-vue/lib/textarea/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _typeof = require('@babel/runtime/helpers/typeof');
var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../config.js');
var textarea_props = require('./props.js');
var _common_js_utils_calcTextareaHeight = require('../_common/js/utils/calcTextareaHeight.js');
var form_hooks = require('../form/hooks.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
var shared_useVModel_index = require('../shared/useVModel/index.js');
var shared_util = require('../shared/util.js');
var shared_component = require('../shared/component.js');
require('lodash/isNull');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/toConsumableArray');
require('@babel/runtime/helpers/objectWithoutProperties');
require('lodash/isString');
require('lodash/isUndefined');
require('lodash/isNumber');
require('lodash/isArray');
require('../config-provider/useConfig.js');
require('lodash/isFunction');
require('lodash/cloneDeep');
require('../config-provider/context.js');
require('lodash/mergeWith');
require('lodash/merge');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-191173f8.js');
require('../_chunks/dep-2088457f.js');
require('dayjs');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../hooks/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isObject');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _typeof__default = /*#__PURE__*/_interopDefaultLegacy(_typeof);
var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var name = "".concat(prefix, "-textarea");
var Textarea = vue.defineComponent({
  name: name,
  props: textarea_props["default"],
  setup: function setup(props, context) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var disabled = form_hooks.useFormDisabled();
    var textareaClass = hooks_useClass.usePrefixClass("textarea");
    var textareaClasses = vue.computed(function () {
      return ["".concat(textareaClass.value), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(textareaClass.value, "--layout-").concat(props.layout), props.layout), "".concat(textareaClass.value, "--border"), props.bordered)];
    });
    var textareaInnerClasses = vue.computed(function () {
      return ["".concat(textareaClass.value, "__wrapper-inner"), _defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(textareaClass.value, "--disabled"), disabled.value), "".concat(textareaClass.value, "--readonly"), props.readonly)];
    });
    var textareaRef = vue.ref();
    var textareaStyle = vue.ref();
    var textareaLength = vue.ref(0);
    var _toRefs = vue.toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = shared_useVModel_index.useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray__default["default"](_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var setInputValue = function setInputValue() {
      var v = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var input = textareaRef.value;
      var sV = String(v);
      if (!input) {
        return;
      }
      if (input.value !== sV) {
        input.value = sV;
      }
    };
    var adjustTextareaHeight = function adjustTextareaHeight() {
      if (props.autosize === true) {
        textareaStyle.value = _common_js_utils_calcTextareaHeight["default"](textareaRef.value);
      } else if (props.autosize === false) {
        textareaStyle.value = _common_js_utils_calcTextareaHeight["default"](textareaRef.value, 1, 1);
      } else if (_typeof__default["default"](props.autosize) === "object") {
        var _props$autosize = props.autosize,
          minRows = _props$autosize.minRows,
          maxRows = _props$autosize.maxRows;
        textareaStyle.value = _common_js_utils_calcTextareaHeight["default"](textareaRef.value, minRows, maxRows);
      } else if (context.attrs.rows) {
        textareaStyle.value = {
          height: "auto",
          minHeight: "auto"
        };
      }
    };
    var handleInput = function handleInput(e) {
      if (e instanceof InputEvent) {
        if (e.isComposing || e.inputType === "insertCompositionText") return;
      }
      textareaValueChangeHandle();
    };
    var textareaValueChangeHandle = function textareaValueChangeHandle() {
      var textarea = textareaRef.value;
      if (!props.allowInputOverMax && props.maxcharacter && props.maxcharacter > 0 && !Number.isNaN(props.maxcharacter)) {
        var _getCharacterLength = shared_util.getCharacterLength(textarea.value, props.maxcharacter),
          _getCharacterLength$l = _getCharacterLength.length,
          length = _getCharacterLength$l === void 0 ? 0 : _getCharacterLength$l,
          _getCharacterLength$c = _getCharacterLength.characters,
          characters = _getCharacterLength$c === void 0 ? "" : _getCharacterLength$c;
        setInnerValue(characters);
        textareaLength.value = length;
      } else {
        setInnerValue(textarea.value);
        textareaLength.value = textarea.value.length;
      }
      vue.nextTick(function () {
        return setInputValue(innerValue.value);
      });
      adjustTextareaHeight();
    };
    var handleCompositionend = function handleCompositionend(e) {
      textareaValueChangeHandle();
    };
    var handleFocus = function handleFocus(e) {
      var _props$onFocus;
      (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, innerValue.value, {
        e: e
      });
    };
    var handleBlur = function handleBlur(e) {
      var _props$onBlur;
      (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, innerValue.value, {
        e: e
      });
    };
    vue.onMounted(function () {
      if (props.autofocus) {
        var _textareaRef$value;
        (_textareaRef$value = textareaRef.value) === null || _textareaRef$value === void 0 || _textareaRef$value.focus();
      }
      textareaValueChangeHandle();
      adjustTextareaHeight();
    });
    vue.watch(innerValue, function () {
      vue.nextTick(function () {
        adjustTextareaHeight();
      });
    });
    vue.watch(function () {
      return props.autosize;
    }, function () {
      adjustTextareaHeight();
    });
    return function () {
      var renaderLabel = function renaderLabel() {
        var label = renderTNodeJSX("label");
        return label ? vue.createVNode("div", {
          "class": "".concat(textareaClass.value, "__label")
        }, [label]) : null;
      };
      var readerIndicator = function readerIndicator() {
        var isShowIndicator = props.indicator && (props.maxcharacter || props.maxlength);
        if (!isShowIndicator) {
          return null;
        }
        return vue.createVNode("div", {
          "class": "".concat(textareaClass.value, "__indicator")
        }, ["".concat(textareaLength.value, "/").concat(props.maxcharacter || props.maxlength)]);
      };
      return vue.createVNode("div", {
        "class": textareaClasses.value
      }, [renaderLabel(), vue.createVNode("div", {
        "class": "".concat(textareaClass.value, "__wrapper")
      }, [vue.createVNode("textarea", {
        "ref": textareaRef,
        "value": innerValue.value,
        "class": textareaInnerClasses.value,
        "style": textareaStyle.value,
        "name": props.name,
        "maxlength": props.maxlength,
        "disabled": props.disabled,
        "placeholder": props.placeholder,
        "readonly": props.readonly,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onCompositionend": handleCompositionend
      }, null), readerIndicator()])]);
    };
  }
});

var _Textarea = shared_component.withInstall(Textarea);

exports["default"] = _Textarea;
