// tdesign-mobile-vue/es/skeleton/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, ref, createVNode, h, watch } from 'vue';
import isNumber from 'lodash/isNumber';
import isArray from 'lodash/isArray';
import isFunction from 'lodash/isFunction';
import { useContent } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import config from '../config.js';
import SkeletonProps from './props.js';
import '../shared/index.js';
import './style/css.js';
import './type.js';
import { withInstall } from '../shared/component.js';
import 'lodash/camelCase';
import 'lodash/kebabCase';
import '../hooks/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isString';
import 'lodash/isObject';
import '../config-provider/useConfig.js';
import 'lodash/cloneDeep';
import '../config-provider/context.js';
import 'lodash/mergeWith';
import 'lodash/merge';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-4c6045f6.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-1d280204.js';
import 'dayjs';
import '../_chunks/dep-5101c61a.js';
import '../config-provider/type.js';
import '../shared/functions.js';
import '../shared/util.js';
import '@babel/runtime/helpers/slicedToArray';
import '../shared/constants.js';
import '../shared/render.js';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useVModel/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '../shared/useGesture/index.js';
import '@use-gesture/vanilla';
import '../shared/hover.js';

var prefix = config.prefix;
var name = "".concat(prefix, "-skeleton");
var ThemeMap = {
  avatar: [{
    type: "circle",
    size: "48px"
  }],
  image: [{
    type: "rect",
    size: "72px"
  }],
  text: [[{
    width: "24%",
    height: "16px",
    marginRight: "16px"
  }, {
    width: "76%",
    height: "16px"
  }], 1],
  paragraph: [1, 1, 1, {
    width: "55%"
  }]
};
var getColItemStyle = function getColItemStyle(obj) {
  var styleName = ["width", "height", "marginRight", "marginLeft", "margin", "size", "background", "backgroundColor", "borderRadius"];
  var style = {};
  styleName.forEach(function (name2) {
    if (name2 in obj) {
      var px = isNumber(obj[name2]) ? "".concat(obj[name2], "px") : obj[name2];
      if (name2 === "size") {
        var _ref = [px, px];
        style.width = _ref[0];
        style.height = _ref[1];
      } else {
        style[name2] = px;
      }
    }
  });
  return style;
};
var Skeleton = defineComponent({
  name: name,
  props: SkeletonProps,
  setup: function setup(props, _ref2) {
    var slots = _ref2.slots;
    var isShow = ref(false);
    var renderContent = useContent();
    var componentName = usePrefixClass("skeleton");
    var getColItemClass = function getColItemClass(obj) {
      return ["".concat(componentName.value, "__col"), "".concat(componentName.value, "--type-").concat(obj.type || "text"), _defineProperty({}, "".concat(componentName.value, "--animation-").concat(props.animation), props.animation)];
    };
    var renderCols = function renderCols(_cols) {
      var cols = [];
      if (isArray(_cols)) {
        cols = _cols;
      } else if (isNumber(_cols)) {
        cols = new Array(_cols).fill({
          type: "text"
        });
      } else {
        cols = [_cols];
      }
      return cols.map(function (obj) {
        return createVNode("div", {
          "class": getColItemClass(obj),
          "style": getColItemStyle(obj)
        }, [isFunction(obj.content) ? obj.content(h) : obj.content]);
      });
    };
    var renderRowCol = function renderRowCol(_rowCol) {
      var rowCol = _rowCol || props.rowCol;
      return rowCol.map(function (item) {
        return createVNode("div", {
          "class": "".concat(name, "__row")
        }, [renderCols(item)]);
      });
    };
    watch(function () {
      return props.loading;
    }, function (val) {
      if (!val || props.delay === 0) {
        isShow.value = val;
        return;
      }
      setTimeout(function () {
        isShow.value = val;
      }, props.delay);
    }, {
      immediate: true
    });
    return function () {
      var content = renderContent("default", "content");
      if (slots.default) {
        return createVNode("div", null, [content]);
      }
      if (!isShow.value) {
        return;
      }
      var children = [];
      if (props.rowCol) {
        children.push(renderRowCol(props.rowCol));
      } else if (props.theme) {
        children.push(renderRowCol(ThemeMap[props.theme]));
      } else if (!props.theme && !props.rowCol) {
        children.push(renderRowCol([[{
          width: "24%",
          height: "16px",
          marginRight: "16px"
        }, {
          width: "76%",
          height: "16px"
        }], 1]));
      }
      return createVNode("div", {
        "class": componentName.value
      }, [children]);
    };
  }
});

var _Skeleton = withInstall(Skeleton);

export { _Skeleton as default };
