// tdesign-mobile-vue/es/avatar/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, inject, ref, computed, createVNode, mergeProps, provide, Fragment } from 'vue';
import _Badge from '../badge/index.js';
import _Image from '../image/index.js';
import config from '../config.js';
import AvatarProps from './props.js';
import ClASSNAMES from '../shared/constants.js';
import { useTNodeJSX, useContent } from '../hooks/tnode.js';
import { usePrefixClass } from '../hooks/useClass.js';
import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import AvatarGroupProps from './avatar-group-props.js';
import '../shared/index.js';
import './style/css.js';
import './type.js';
import { withInstall } from '../shared/component.js';
import '@babel/runtime/helpers/typeof';
import '@babel/runtime/helpers/slicedToArray';
import '../badge/props.js';
import '../badge/type.js';
import '../config-provider/useConfig.js';
import 'lodash/isFunction';
import 'lodash/cloneDeep';
import 'lodash/isString';
import '../config-provider/context.js';
import 'lodash/mergeWith';
import 'lodash/merge';
import 'lodash/isArray';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-4c6045f6.js';
import '../_chunks/dep-1d280204.js';
import 'dayjs';
import '../_chunks/dep-5101c61a.js';
import '../config-provider/type.js';
import 'lodash/camelCase';
import 'lodash/kebabCase';
import '../hooks/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isObject';
import '../shared/functions.js';
import '../shared/util.js';
import 'lodash/isNumber';
import '../shared/render.js';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import '../shared/useChildSlots/index.js';
import '../shared/useVModel/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import '../shared/useGesture/index.js';
import '@use-gesture/vanilla';
import '../shared/hover.js';
import '../image/image.js';
import 'tdesign-icons-vue-next';
import '../loading/index.js';
import '../_common/js/loading/circle-adapter.js';
import '../_common/js/utils/set-style.js';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import 'lodash/isNull';
import 'lodash/isUndefined';
import '../loading/props.js';
import '../loading/type.js';
import '../image/props.js';
import '../image/type.js';

var prefix$1 = config.prefix;
var name$1 = "".concat(prefix$1, "-avatar");
var _Avatar = defineComponent({
  name: name$1,
  props: AvatarProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var renderTNodeContent = useContent();
    var avatarClass = usePrefixClass("avatar");
    var avatarGroupProps = inject("avatarGroup", {});
    var hasAvatarGroupProps = Object.keys(avatarGroupProps).length > 0;
    var sizeValue = ref(props.size || avatarGroupProps && avatarGroupProps.size);
    var sizeReValue = ref(avatarGroupProps && avatarGroupProps.size || props.size);
    var sizeClass = "".concat(sizeReValue.value.indexOf("px") > -1 ? "medium" : sizeReValue.value);
    var avatarClasses = computed(function () {
      return ["".concat(avatarClass.value), "".concat(avatarClass.value, "--").concat(sizeClass), _defineProperty({}, "".concat(avatarClass.value, "--").concat(props.shape), props.shape), hasAvatarGroupProps ? "".concat(avatarClass.value, "--border ").concat(avatarClass.value, "--border-").concat(sizeClass) : ""];
    });
    var isCustomSize = computed(function () {
      return sizeValue.value && !ClASSNAMES.SIZE[sizeValue.value];
    });
    var customSize = computed(function () {
      return isCustomSize.value ? {
        height: sizeValue.value,
        width: sizeValue.value
      } : {};
    });
    var handleImgLoadError = function handleImgLoadError(e) {
      var _props$onError;
      (_props$onError = props.onError) === null || _props$onError === void 0 || _props$onError.call(props, e);
    };
    return function () {
      var icon = renderTNodeJSX("icon");
      var TNodeContent = renderTNodeContent("default", "content");
      var readerAvatar = function readerAvatar() {
        if (props.image && !props.hideOnLoadFailed) {
          return createVNode(_Image, mergeProps({
            "style": customSize.value,
            "src": props.image,
            "alt": props.alt
          }, props.imageProps, {
            "onError": handleImgLoadError
          }), null);
        }
        if (icon) {
          return createVNode("div", {
            "class": "".concat(avatarClass.value, "__icon")
          }, [icon]);
        }
        return createVNode("span", null, [TNodeContent]);
      };
      return createVNode("div", {
        "class": "".concat(avatarClass.value, "__wrapper"),
        "style": customSize.value
      }, [createVNode("div", {
        "class": "".concat(avatarClass.value, "__badge")
      }, [createVNode(_Badge, props.badgeProps, {
        default: function _default() {
          return [createVNode("div", {
            "class": avatarClasses.value
          }, [readerAvatar()])];
        }
      })])]);
    };
  }
});

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config.prefix;
var name = "".concat(prefix, "-avatar-group");
var _AvatarGroup = defineComponent({
  name: name,
  props: AvatarGroupProps,
  setup: function setup(props) {
    var renderTNodeJSX = useTNodeJSX();
    var avatarGroupClass = usePrefixClass("avatar-group");
    provide("avatarGroup", _objectSpread({}, props));
    var direction = props.cascading ? props.cascading.split("-")[0] : "right";
    var avatarGroupClasses = computed(function () {
      return ["".concat(avatarGroupClass.value), "".concat(avatarGroupClass.value, "-offset-").concat(direction, "-").concat(props.size.indexOf("px") > -1 ? "medium" : props.size)];
    });
    var readerAvatar = function readerAvatar() {
      var children = renderTNodeJSX("default");
      var allChildren = [];
      children.forEach(function (child) {
        if (child.type === Fragment) {
          allChildren.push.apply(allChildren, _toConsumableArray(child.children));
        } else {
          allChildren.push(child);
        }
      });
      var isShowCollapse = false;
      var avatarList = [];
      if (allChildren.length > props.max) {
        avatarList = allChildren.slice(0, props.max);
        isShowCollapse = true;
      } else {
        avatarList = allChildren;
      }
      if (props.cascading === "left-up") {
        var defaultZIndex = 100;
        for (var index = 0; index < avatarList.length; index++) {
          avatarList[index].props.style = "z-index: ".concat(defaultZIndex - index * 10);
        }
      }
      if (isShowCollapse) {
        var collapseAvatar = renderTNodeJSX("collapseAvatar");
        avatarList.push(createVNode(_Avatar, {
          "size": avatarList[0].size || props.size
        }, {
          default: function _default() {
            return [collapseAvatar || "+".concat(allChildren.length - props.max)];
          }
        }));
      }
      return avatarList;
    };
    return function () {
      return createVNode("div", {
        "class": avatarGroupClasses.value
      }, [readerAvatar()]);
    };
  }
});

var Avatar = withInstall(_Avatar);
var AvatarGroup = withInstall(_AvatarGroup);

export { AvatarGroup, Avatar as default };
