// tdesign-mobile-vue/esm/textarea/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import '../shared/index.js';
import _typeof from '@babel/runtime/helpers/typeof';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, computed, ref, toRefs, nextTick, onMounted, watch, createVNode } from 'vue';
import config from '../config.js';
import TextareaProps from './props.js';
import calcTextareaHeight from '../_common/js/utils/calcTextareaHeight.js';
import { useFormDisabled } from '../form/hooks.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useTNodeJSX } from '../hooks/tnode.js';
import { useVModel } from '../shared/useVModel/index.js';
import { getCharacterLength } from '../shared/util.js';
import './style';
import './type.js';
import { withInstall } from '../shared/component.js';
import '../shared/functions.js';
import '../shared/constants.js';
import '../shared/render.js';
import 'lodash/camelCase';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import '../shared/useChildSlots/index.js';
import '@babel/runtime/helpers/toConsumableArray';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import 'lodash/isArray';
import '../shared/useGesture/index.js';
import '@use-gesture/vanilla';
import '../shared/hover.js';
import 'lodash/isNull';
import '../_common/js/utils/helper.js';
import '@babel/runtime/helpers/objectWithoutProperties';
import 'lodash/isString';
import 'lodash/isUndefined';
import 'lodash/isNumber';
import '../config-provider/useConfig.js';
import 'lodash/isFunction';
import 'lodash/cloneDeep';
import '../config-provider/context.js';
import 'lodash/mergeWith';
import 'lodash/merge';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-4c6045f6.js';
import '../_chunks/dep-1d280204.js';
import 'dayjs';
import '../_chunks/dep-5101c61a.js';
import '../config-provider/type.js';
import 'lodash/kebabCase';
import '../hooks/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isObject';

var prefix = config.prefix;
var name = "".concat(prefix, "-textarea");
var Textarea = defineComponent({
  name: name,
  props: TextareaProps,
  setup: function setup(props, context) {
    var renderTNodeJSX = useTNodeJSX();
    var disabled = useFormDisabled();
    var textareaClass = usePrefixClass("textarea");
    var textareaClasses = computed(function () {
      return ["".concat(textareaClass.value), _defineProperty(_defineProperty({}, "".concat(textareaClass.value, "--layout-").concat(props.layout), props.layout), "".concat(textareaClass.value, "--border"), props.bordered)];
    });
    var textareaInnerClasses = computed(function () {
      return ["".concat(textareaClass.value, "__wrapper-inner"), _defineProperty(_defineProperty({}, "".concat(textareaClass.value, "--disabled"), disabled.value), "".concat(textareaClass.value, "--readonly"), props.readonly)];
    });
    var textareaRef = ref();
    var textareaStyle = ref();
    var textareaLength = ref(0);
    var _toRefs = toRefs(props),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props.defaultValue, props.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      innerValue = _useVModel2[0],
      setInnerValue = _useVModel2[1];
    var setInputValue = function setInputValue() {
      var v = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var input = textareaRef.value;
      var sV = String(v);
      if (!input) {
        return;
      }
      if (input.value !== sV) {
        input.value = sV;
      }
    };
    var adjustTextareaHeight = function adjustTextareaHeight() {
      if (props.autosize === true) {
        textareaStyle.value = calcTextareaHeight(textareaRef.value);
      } else if (props.autosize === false) {
        textareaStyle.value = calcTextareaHeight(textareaRef.value, 1, 1);
      } else if (_typeof(props.autosize) === "object") {
        var _props$autosize = props.autosize,
          minRows = _props$autosize.minRows,
          maxRows = _props$autosize.maxRows;
        textareaStyle.value = calcTextareaHeight(textareaRef.value, minRows, maxRows);
      } else if (context.attrs.rows) {
        textareaStyle.value = {
          height: "auto",
          minHeight: "auto"
        };
      }
    };
    var handleInput = function handleInput(e) {
      if (e instanceof InputEvent) {
        if (e.isComposing || e.inputType === "insertCompositionText") return;
      }
      textareaValueChangeHandle();
    };
    var textareaValueChangeHandle = function textareaValueChangeHandle() {
      var textarea = textareaRef.value;
      if (!props.allowInputOverMax && props.maxcharacter && props.maxcharacter > 0 && !Number.isNaN(props.maxcharacter)) {
        var _getCharacterLength = getCharacterLength(textarea.value, props.maxcharacter),
          _getCharacterLength$l = _getCharacterLength.length,
          length = _getCharacterLength$l === void 0 ? 0 : _getCharacterLength$l,
          _getCharacterLength$c = _getCharacterLength.characters,
          characters = _getCharacterLength$c === void 0 ? "" : _getCharacterLength$c;
        setInnerValue(characters);
        textareaLength.value = length;
      } else {
        setInnerValue(textarea.value);
        textareaLength.value = textarea.value.length;
      }
      nextTick(function () {
        return setInputValue(innerValue.value);
      });
      adjustTextareaHeight();
    };
    var handleCompositionend = function handleCompositionend(e) {
      textareaValueChangeHandle();
    };
    var handleFocus = function handleFocus(e) {
      var _props$onFocus;
      (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, innerValue.value, {
        e: e
      });
    };
    var handleBlur = function handleBlur(e) {
      var _props$onBlur;
      (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, innerValue.value, {
        e: e
      });
    };
    onMounted(function () {
      if (props.autofocus) {
        var _textareaRef$value;
        (_textareaRef$value = textareaRef.value) === null || _textareaRef$value === void 0 || _textareaRef$value.focus();
      }
      textareaValueChangeHandle();
      adjustTextareaHeight();
    });
    watch(innerValue, function () {
      nextTick(function () {
        adjustTextareaHeight();
      });
    });
    watch(function () {
      return props.autosize;
    }, function () {
      adjustTextareaHeight();
    });
    return function () {
      var renaderLabel = function renaderLabel() {
        var label = renderTNodeJSX("label");
        return label ? createVNode("div", {
          "class": "".concat(textareaClass.value, "__label")
        }, [label]) : null;
      };
      var readerIndicator = function readerIndicator() {
        var isShowIndicator = props.indicator && (props.maxcharacter || props.maxlength);
        if (!isShowIndicator) {
          return null;
        }
        return createVNode("div", {
          "class": "".concat(textareaClass.value, "__indicator")
        }, ["".concat(textareaLength.value, "/").concat(props.maxcharacter || props.maxlength)]);
      };
      return createVNode("div", {
        "class": textareaClasses.value
      }, [renaderLabel(), createVNode("div", {
        "class": "".concat(textareaClass.value, "__wrapper")
      }, [createVNode("textarea", {
        "ref": textareaRef,
        "value": innerValue.value,
        "class": textareaInnerClasses.value,
        "style": textareaStyle.value,
        "name": props.name,
        "maxlength": props.maxlength,
        "disabled": props.disabled,
        "placeholder": props.placeholder,
        "readonly": props.readonly,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onCompositionend": handleCompositionend
      }, null), readerIndicator()])]);
    };
  }
});

var _Textarea = withInstall(Textarea);

export { _Textarea as default };
