// tdesign-mobile-vue/lib/_chunks/dep-77f965a9.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var popup_index = require('../popup/index.js');
var config = require('../config.js');
var drawer_props = require('../drawer/props.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var shared_util = require('../shared/util.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var name = "".concat(prefix, "-drawer");
var _Drawer = vue.defineComponent({
  name: name,
  components: {
    TPopup: popup_index["default"]
  },
  props: drawer_props["default"],
  emits: ["update:visible", "itemClick", "overlayClick"],
  setup: function setup(props2, context) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var drawerClass = hooks_useClass.usePrefixClass("drawer");
    var _toRefs = vue.toRefs(props2),
      visible = _toRefs.visible,
      showOverlay = _toRefs.showOverlay;
    var open = vue.ref(visible.value || false);
    vue.watch(open, function () {
      context.emit("update:visible", open.value);
    });
    vue.watch(visible, function () {
      open.value = visible.value;
    });
    var onItemClick = function onItemClick(index, item, context2) {
      var _props2$onItemClick;
      (_props2$onItemClick = props2.onItemClick) === null || _props2$onItemClick === void 0 || _props2$onItemClick.call(props2, index, item, context2);
    };
    var onVisibleChange = function onVisibleChange(visible2) {
      if (showOverlay.value) {
        var _props2$onOverlayClic;
        (_props2$onOverlayClic = props2.onOverlayClick) === null || _props2$onOverlayClic === void 0 || _props2$onOverlayClic.call(props2, {
          visible: visible2
        });
      }
    };
    var onClose = function onClose() {
      var _props2$onClose;
      (_props2$onClose = props2.onClose) === null || _props2$onClose === void 0 || _props2$onClose.call(props2, "overlay");
    };
    return function () {
      var attach = props2.attach,
        zIndex = props2.zIndex,
        closeOnOverlayClick = props2.closeOnOverlayClick,
        destroyOnClose = props2.destroyOnClose,
        items = props2.items,
        placement = props2.placement,
        showOverlay2 = props2.showOverlay;
      var renderTitleNode = function renderTitleNode() {
        var titleNode = renderTNodeJSX("title");
        if (!titleNode) {
          return null;
        }
        return vue.createVNode("div", {
          "class": "".concat(drawerClass.value, "__title")
        }, [titleNode]);
      };
      var renderFooterNode = function renderFooterNode() {
        var footerNode = renderTNodeJSX("footer");
        if (!footerNode) {
          return null;
        }
        return vue.createVNode("div", {
          "class": "".concat(drawerClass.value, "__footer")
        }, [footerNode]);
      };
      return vue.createVNode(vue.resolveComponent("t-popup"), {
        "modelValue": open.value,
        "onUpdate:modelValue": function onUpdateModelValue($event) {
          return open.value = $event;
        },
        "placement": placement,
        "attach": attach,
        "showOverlay": showOverlay2,
        "zIndex": zIndex,
        "closeOnOverlayClick": closeOnOverlayClick,
        "destroyOnClose": destroyOnClose,
        "onVisibleChange": onVisibleChange,
        "onClose": onClose
      }, {
        default: function _default() {
          return [vue.createVNode("div", {
            "class": "".concat(drawerClass.value)
          }, [renderTitleNode(), renderTNodeJSX("default"), vue.createVNode("div", {
            "class": "".concat(drawerClass.value, "__sidebar")
          }, [(items || []).map(function (item, index) {
            return vue.createVNode("div", {
              "key": item.title,
              "class": "".concat(drawerClass.value, "__sidebar-item"),
              "onClick": function onClick(e) {
                return onItemClick(index, item, {
                  e: e
                });
              }
            }, [item.icon && vue.createVNode("span", {
              "class": "".concat(drawerClass.value, "__sidebar-item-icon")
            }, [item.icon(vue.h)]), vue.createVNode("div", {
              "class": "".concat(drawerClass.value, "__sidebar-item-title")
            }, [item.title])]);
          })]), renderFooterNode()])];
        }
      });
    };
  }
});

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var Drawer = function Drawer(options) {
  if (!shared_util.isBrowser) return;
  var root = document.createElement("div");
  document.body.appendChild(root);
  var visible = vue.ref(false);
  var props = vue.ref({});
  var destroyOnClose = vue.ref(false);
  vue.createApp(function () {
    return vue.h(_Drawer, _objectSpread(_objectSpread({}, options), {}, {
      visible: visible.value,
      destroyOnClose: destroyOnClose.value
    }, props.value));
  }).mount(root);
  var handler = {
    destroy: function destroy() {
      destroyOnClose.value = true;
      vue.nextTick(function () {
        visible.value = false;
        document.body.removeChild(root);
      });
    },
    hide: function hide() {
      visible.value = false;
    },
    show: function show() {
      visible.value = true;
    },
    update: function update(options2) {
      props.value = options2;
    }
  };
  return handler;
};
Drawer.install = function (app) {
  app.config.globalProperties.$drawer = Drawer;
};
var DrawerPlugin = Drawer;

exports.DrawerPlugin = DrawerPlugin;
exports._Drawer = _Drawer;
