// tdesign-mobile-vue/lib/_chunks/dep-d44af455.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

var button_index = require('../button/index.js');
var divider_index = require('../divider/index.js');
var calendar_index = require('../calendar/index.js');
var cell_index = require('../cell/index.js');
var checkbox_index = require('../checkbox/index.js');
var input_index = require('../input/index.js');
var textarea_index = require('../textarea/index.js');
var overlay_index = require('../overlay/index.js');
var progress_index = require('../progress/index.js');
var radio_index = require('../radio/index.js');
var rate_index = require('../rate/index.js');
var switch_index = require('../switch/index.js');
var stepper_index = require('../stepper/index.js');
var slider_index = require('../slider/index.js');
var dialog_index = require('../dialog/index.js');
var popup_index = require('../popup/index.js');
var actionSheet_index = require('../action-sheet/index.js');
var badge_index = require('../badge/index.js');
var fab_index = require('../fab/index.js');
var imageViewer_index = require('../image-viewer/index.js');
var drawer_index = require('../drawer/index.js');
var navbar_index = require('../navbar/index.js');
var message_index = require('../message/index.js');
var dateTimePicker_index = require('../date-time-picker/index.js');
var toast_index = require('../toast/index.js');
var search_index = require('../search/index.js');
var steps_index = require('../steps/index.js');
var indexes_index = require('../indexes/index.js');
var picker_index = require('../picker/index.js');
var tag_index = require('../tag/index.js');
var tabBar_index = require('../tab-bar/index.js');
var dropdownMenu_index = require('../dropdown-menu/index.js');
var collapse_index = require('../collapse/index.js');
var tabs_index = require('../tabs/index.js');
var swiper_index = require('../swiper/index.js');
var countDown_index = require('../count-down/index.js');
var avatar_index = require('../avatar/index.js');
var image_index = require('../image/index.js');
var swipeCell_index = require('../swipe-cell/index.js');
var skeleton_index = require('../skeleton/index.js');
var grid_index = require('../grid/index.js');
var sticky_index = require('../sticky/index.js');
var list_index = require('../list/index.js');
var loading_index = require('../loading/index.js');
var backTop_index = require('../back-top/index.js');
var pullDownRefresh_index = require('../pull-down-refresh/index.js');
var upload_index = require('../upload/index.js');
var noticeBar_index = require('../notice-bar/index.js');
var result_index = require('../result/index.js');
var cascader_index = require('../cascader/index.js');
var footer_index = require('../footer/index.js');
var empty_index = require('../empty/index.js');
var link_index = require('../link/index.js');
var form_index = require('../form/index.js');
var icon_index = require('../icon/index.js');
var table_index = require('../table/index.js');
var sideBar_index = require('../side-bar/index.js');
var popover_index = require('../popover/index.js');
var guide_index = require('../guide/index.js');
var configProvider_index = require('../config-provider/index.js');

var components = /*#__PURE__*/Object.freeze({
	__proto__: null,
	Button: button_index["default"],
	Divider: divider_index["default"],
	Calendar: calendar_index["default"],
	Cell: cell_index["default"],
	CellGroup: cell_index.CellGroup,
	Checkbox: checkbox_index["default"],
	CheckboxGroup: checkbox_index.CheckboxGroup,
	Input: input_index["default"],
	Textarea: textarea_index["default"],
	Overlay: overlay_index["default"],
	Progress: progress_index["default"],
	Radio: radio_index["default"],
	RadioGroup: radio_index.RadioGroup,
	Rate: rate_index["default"],
	Switch: switch_index["default"],
	Stepper: stepper_index["default"],
	Slider: slider_index["default"],
	Dialog: dialog_index.DialogPlugin,
	Popup: popup_index["default"],
	ActionSheet: actionSheet_index["default"],
	Badge: badge_index["default"],
	Fab: fab_index["default"],
	ImageViewer: imageViewer_index["default"],
	Drawer: drawer_index["default"],
	Navbar: navbar_index["default"],
	Message: message_index.MessagePlugin,
	DateTimePicker: dateTimePicker_index._DateTimePicker,
	Toast: toast_index.ToastPlugin,
	Search: search_index["default"],
	Steps: steps_index.Steps,
	StepItem: steps_index.StepItem,
	Indexes: indexes_index.Indexes,
	IndexesAnchor: indexes_index.IndexesAnchor,
	Picker: picker_index.Picker,
	Cascade: picker_index.Cascade,
	Tag: tag_index.Tag,
	CheckTag: tag_index.CheckTag,
	TabBar: tabBar_index.TabBar,
	TabBarItem: tabBar_index.TabBarItem,
	DropdownMenu: dropdownMenu_index.DropdownMenu,
	DropdownItem: dropdownMenu_index.DropdownItem,
	Collapse: collapse_index.Collapse,
	CollapsePanel: collapse_index.CollapsePanel,
	Tabs: tabs_index.Tabs,
	TabPanel: tabs_index.TabPanel,
	Swiper: swiper_index.Swiper,
	SwiperItem: swiper_index.SwiperItem,
	CountDown: countDown_index.CountDown,
	Avatar: avatar_index["default"],
	AvatarGroup: avatar_index.AvatarGroup,
	Image: image_index["default"],
	SwipeCell: swipeCell_index["default"],
	Skeleton: skeleton_index["default"],
	Grid: grid_index.Grid,
	GridItem: grid_index.GridItem,
	Sticky: sticky_index["default"],
	List: list_index["default"],
	Loading: loading_index["default"],
	BackTop: backTop_index["default"],
	PullDownRefresh: pullDownRefresh_index["default"],
	Upload: upload_index["default"],
	NoticeBar: noticeBar_index["default"],
	Result: result_index["default"],
	Cascader: cascader_index["default"],
	Footer: footer_index["default"],
	Empty: empty_index["default"],
	Link: link_index["default"],
	Form: form_index.Form,
	FormItem: form_index.FormItem,
	Icon: icon_index.Icon,
	Table: table_index["default"],
	SideBar: sideBar_index.SideBar,
	SideBarItem: sideBar_index.SideBarItem,
	TPopover: popover_index["default"],
	Guide: guide_index["default"],
	ConfigProvider: configProvider_index.ConfigProvider
});

exports.components = components;
