// tdesign-mobile-vue/es/config-provider/config-provider.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import { defineComponent, renderSlot } from 'vue';
import config from '../config.js';
import { provideConfig } from './useConfig.js';
import 'lodash/isFunction';
import 'lodash/cloneDeep';
import 'lodash/isString';
import './context.js';
import 'lodash/mergeWith';
import 'lodash/merge';
import 'lodash/isArray';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-4c6045f6.js';
import '@babel/runtime/helpers/typeof';
import '../_chunks/dep-1d280204.js';
import 'dayjs';
import '../_chunks/dep-5101c61a.js';
import './type.js';

var prefix = config.prefix;
var name = "".concat(prefix, "-config-provider");
var configProviderProps = {
  globalConfig: Object
};
var script = defineComponent({
  name: name,
  props: configProviderProps,
  setup: function setup(props) {
    provideConfig(props);
    return {
      name: name
    };
  }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  return renderSlot(_ctx.$slots, "default");
}

script.render = render;

export { configProviderProps, script as default };
