// tdesign-mobile-vue/lib/input/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _slicedToArray = require('@babel/runtime/helpers/slicedToArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var core = require('@vueuse/core');
var config = require('../config.js');
var input_props = require('./props.js');
var form_hooks = require('../form/hooks.js');
var hooks_useClass = require('../hooks/useClass.js');
var hooks_tnode = require('../hooks/tnode.js');
var shared_component = require('../shared/component.js');
var shared_util = require('../shared/util.js');
var shared_useDefault_index = require('../shared/useDefault/index.js');
require('../config-provider/useConfig.js');
require('lodash/isFunction');
require('lodash/cloneDeep');
require('lodash/isString');
require('../config-provider/context.js');
require('lodash/mergeWith');
require('lodash/merge');
require('lodash/isArray');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-191173f8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2088457f.js');
require('dayjs');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../hooks/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isObject');
require('lodash/isNumber');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _slicedToArray__default = /*#__PURE__*/_interopDefaultLegacy(_slicedToArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var name = "".concat(prefix, "-input");
var Input = vue.defineComponent({
  name: name,
  props: _objectSpread(_objectSpread({}, input_props["default"]), {}, {
    labelAlign: {
      type: String,
      default: "top"
    },
    pattern: {
      type: String
    },
    inputmode: {
      type: String,
      validator: function validator(val) {
        if (!val) return true;
        return ["search", "text", "none", "url", "tel", "email", "numeric", "decimal"].includes(val);
      }
    }
  }),
  setup: function setup(props, context) {
    var readerTNodeJSX = hooks_tnode.useTNodeJSX();
    var inputClass = hooks_useClass.usePrefixClass("input");
    var isDisabled = form_hooks.useFormDisabled();
    var inputRef = vue.ref();
    var _toRefs = vue.toRefs(props),
      autofocus = _toRefs.autofocus;
    var _useDefault = shared_useDefault_index.useDefault(props, context.emit, "value", "change"),
      _useDefault2 = _slicedToArray__default["default"](_useDefault, 1),
      innerValue = _useDefault2[0];
    var status = props.status || "default";
    var _useFocus = core.useFocus(inputRef, {
        initialValue: props.autofocus
      }),
      focused = _useFocus.focused;
    var inputClasses = vue.computed(function () {
      return ["".concat(inputClass.value, "__control"), _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(inputClass.value, "--").concat(props.align), props.align !== "left"), "".concat(inputClass.value, "--").concat(status), status), "".concat(inputClass.value, "__control--disabled"), isDisabled.value)];
    });
    var rootClasses = vue.computed(function () {
      return [inputClass.value, _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(inputClass.value, "--layout-").concat(props.layout), props.layout), "".concat(inputClass.value, "--center"), props.labelAlign === "center"), "".concat(inputClass.value, "--border"), !props.borderless)];
    });
    var setInputValue = function setInputValue() {
      var v = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "";
      var input = inputRef.value;
      var sV = String(v);
      if (!input) {
        return;
      }
      if (input.value !== sV) {
        input.value = sV;
      }
    };
    var handleInput = function handleInput(e) {
      if (e instanceof InputEvent) {
        var checkInputType = e.inputType && e.inputType === "insertCompositionText";
        if (e.isComposing || checkInputType) return;
      }
      inputValueChangeHandle(e);
    };
    var inputValueChangeHandle = function inputValueChangeHandle(e) {
      var value = e.target.value;
      var allowInputOverMax = props.allowInputOverMax,
        maxcharacter = props.maxcharacter;
      if (!allowInputOverMax && maxcharacter && maxcharacter > 0 && !Number.isNaN(maxcharacter)) {
        var _getCharacterLength = shared_util.getCharacterLength(value, maxcharacter);
          _getCharacterLength.length;
          var _getCharacterLength$c = _getCharacterLength.characters,
          characters = _getCharacterLength$c === void 0 ? "" : _getCharacterLength$c;
        innerValue.value = characters;
      } else {
        innerValue.value = value;
      }
      vue.nextTick(function () {
        return setInputValue(innerValue.value);
      });
    };
    var focus = function focus() {
      focused.value = true;
    };
    var blur = function blur() {
      focused.value = false;
    };
    shared_component.extendAPI({
      focus: focus,
      blur: blur
    });
    var handleClear = function handleClear(e) {
      var _props$onClear;
      innerValue.value = "";
      focused.value = true;
      (_props$onClear = props.onClear) === null || _props$onClear === void 0 || _props$onClear.call(props, {
        e: e
      });
    };
    var handleFocus = function handleFocus(e) {
      var _props$onFocus;
      (_props$onFocus = props.onFocus) === null || _props$onFocus === void 0 || _props$onFocus.call(props, innerValue.value, {
        e: e
      });
    };
    var handleBlur = function handleBlur(e) {
      var _props$onBlur;
      (_props$onBlur = props.onBlur) === null || _props$onBlur === void 0 || _props$onBlur.call(props, innerValue.value, {
        e: e
      });
    };
    var handleCompositionend = function handleCompositionend(e) {
      inputValueChangeHandle(e);
    };
    vue.watch(autofocus, function (autofocus2, prevAutofocus) {
      if (autofocus2 === true) {
        vue.nextTick(function () {
          focused.value = true;
        });
      }
    });
    return function () {
      var readerPrefix = function readerPrefix() {
        var prefixIcon = readerTNodeJSX("prefixIcon");
        var label = readerTNodeJSX("label");
        return vue.createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--prefix")
        }, [prefixIcon && vue.createVNode("div", {
          "class": "".concat(inputClass.value, "__icon--prefix")
        }, [prefixIcon]), vue.createVNode("div", {
          "class": "".concat(inputClass.value, "__label")
        }, [label])]);
      };
      var readerClearable = function readerClearable() {
        if (props.clearable && innerValue.value && innerValue.value.length > 0) {
          return vue.createVNode("div", {
            "class": "".concat(inputClass.value, "__wrap--clearable-icon"),
            "onClick": handleClear
          }, [vue.createVNode(tdesignIconsVueNext.CloseCircleFilledIcon, null, null)]);
        }
        return null;
      };
      var readerSuffix = function readerSuffix() {
        var suffix = readerTNodeJSX("suffix");
        if (!suffix) {
          return null;
        }
        return vue.createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--suffix")
        }, [suffix]);
      };
      var readerSuffixIcon = function readerSuffixIcon() {
        var suffixIcon = readerTNodeJSX("suffixIcon");
        if (!suffixIcon) {
          return null;
        }
        return vue.createVNode("div", {
          "class": "".concat(inputClass.value, "__wrap--suffix-icon")
        }, [suffixIcon]);
      };
      var readerTips = function readerTips() {
        var tips = readerTNodeJSX("tips");
        if (!tips) {
          return null;
        }
        return vue.createVNode("div", {
          "class": "".concat(inputClass.value, "__tips ").concat(inputClass.value, "--").concat(props.align)
        }, [tips]);
      };
      return vue.createVNode("div", {
        "class": rootClasses.value
      }, [readerPrefix(), vue.createVNode("div", {
        "class": "".concat(inputClass.value, "__wrap")
      }, [vue.createVNode("div", {
        "class": "".concat(inputClass.value, "__content ").concat(inputClass.value, "--").concat(status || "default")
      }, [vue.createVNode("input", {
        "ref": inputRef,
        "value": innerValue.value,
        "name": props.name,
        "class": inputClasses.value,
        "type": props.type,
        "disabled": isDisabled.value,
        "autocomplete": props.autocomplete ? "On" : "Off",
        "placeholder": props.placeholder,
        "readonly": props.readonly,
        "maxlength": props.maxlength || -1,
        "pattern": props.pattern,
        "inputmode": props.inputmode,
        "onFocus": handleFocus,
        "onBlur": handleBlur,
        "onInput": handleInput,
        "onCompositionend": handleCompositionend
      }, null), readerClearable(), readerSuffix(), readerSuffixIcon()]), readerTips()])]);
    };
  }
});

var _Input = shared_component.withInstall(Input);

exports["default"] = _Input;
