// tdesign-mobile-vue/lib/link/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var config = require('../config.js');
var link_props = require('./props.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useClass = require('../hooks/useClass.js');
var shared_component = require('../shared/component.js');
require('lodash/isFunction');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../hooks/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isString');
require('lodash/isObject');
require('../config-provider/useConfig.js');
require('lodash/cloneDeep');
require('../config-provider/context.js');
require('lodash/mergeWith');
require('lodash/merge');
require('lodash/isArray');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-191173f8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2088457f.js');
require('dayjs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

var prefix = config["default"].prefix;
var name = "".concat(prefix, "-link");
var Link = vue.defineComponent({
  name: name,
  props: link_props["default"],
  setup: function setup(props) {
    var linkClass = hooks_useClass.usePrefixClass("link");
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var renderTNodeContent = hooks_tnode.useContent();
    var linkClasses = vue.computed(function () {
      return [linkClass.value, "".concat(linkClass.value, "--").concat(props.theme || "default"), "".concat(linkClass.value, "--").concat(props.size || "medium"), _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(linkClass.value, "--disabled"), props.disabled), "".concat(linkClass.value, "--underline"), props.underline), "".concat(linkClass.value, "--hover"), props.hover && !props.disabled)];
    });
    var handleClick = function handleClick(e) {
      var _props$onClick;
      if (props.disabled) return;
      (_props$onClick = props.onClick) === null || _props$onClick === void 0 || _props$onClick.call(props, e);
    };
    var renderContent = function renderContent() {
      var content = renderTNodeContent("default", "content");
      return content ? vue.createVNode("span", {
        "class": ["".concat(linkClass.value, "__content")]
      }, [content]) : null;
    };
    var renderPrefixIcon = function renderPrefixIcon() {
      var prefixIcon = renderTNodeJSX("prefixIcon");
      return prefixIcon ? vue.createVNode("span", {
        "class": ["".concat(linkClass.value, "__prefix-icon")]
      }, [prefixIcon]) : null;
    };
    var renderSuffixIcon = function renderSuffixIcon() {
      var suffixIcon = renderTNodeJSX("suffixIcon");
      return suffixIcon ? vue.createVNode("span", {
        "class": ["".concat(linkClass.value, "__suffix-icon")]
      }, [suffixIcon]) : null;
    };
    return function () {
      return vue.createVNode("a", {
        "class": linkClasses.value,
        "aria-disabled": props.disabled,
        "target": props.target,
        "href": props.disabled || !props.href ? void 0 : props.href,
        "onClick": handleClick
      }, [renderPrefixIcon(), renderContent(), renderSuffixIcon()]);
    };
  }
});

var _Link = shared_component.withInstall(Link);

exports["default"] = _Link;
