// tdesign-mobile-vue/esm/_chunks/dep-bfec9702.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import _Button from '../button/index.js';
import Divider from '../divider/index.js';
import _Calendar from '../calendar/index.js';
import _Cell, { CellGroup } from '../cell/index.js';
import CheckBox, { CheckboxGroup } from '../checkbox/index.js';
import _Input from '../input/index.js';
import _Textarea from '../textarea/index.js';
import Overlay from '../overlay/index.js';
import _Progress from '../progress/index.js';
import _Radio, { RadioGroup } from '../radio/index.js';
import _Rate from '../rate/index.js';
import _Switch from '../switch/index.js';
import Stepper from '../stepper/index.js';
import _Slider from '../slider/index.js';
import { DialogPlugin } from '../dialog/index.js';
import _Popup from '../popup/index.js';
import _ActionSheet from '../action-sheet/index.js';
import _Badge from '../badge/index.js';
import _Fab from '../fab/index.js';
import _ImageViewer from '../image-viewer/index.js';
import Drawer from '../drawer/index.js';
import _Navbar from '../navbar/index.js';
import { MessagePlugin } from '../message/index.js';
import { _DateTimePicker } from '../date-time-picker/index.js';
import { ToastPlugin } from '../toast/index.js';
import _Search from '../search/index.js';
import { Steps, StepItem } from '../steps/index.js';
import { Indexes, IndexesAnchor } from '../indexes/index.js';
import { Picker, Cascade } from '../picker/index.js';
import { Tag, CheckTag } from '../tag/index.js';
import { TabBar, TabBarItem } from '../tab-bar/index.js';
import { DropdownMenu, DropdownItem } from '../dropdown-menu/index.js';
import { Collapse, CollapsePanel } from '../collapse/index.js';
import { Tabs, TabPanel } from '../tabs/index.js';
import { Swiper, SwiperItem } from '../swiper/index.js';
import { CountDown } from '../count-down/index.js';
import Avatar, { AvatarGroup } from '../avatar/index.js';
import _Image from '../image/index.js';
import SwipeCell from '../swipe-cell/index.js';
import _Skeleton from '../skeleton/index.js';
import { Grid, GridItem } from '../grid/index.js';
import _Sticky from '../sticky/index.js';
import _List from '../list/index.js';
import _Loading from '../loading/index.js';
import _BackTop from '../back-top/index.js';
import _PullDownRefresh from '../pull-down-refresh/index.js';
import _Upload from '../upload/index.js';
import _NoticeBar from '../notice-bar/index.js';
import _Result from '../result/index.js';
import _Cascader from '../cascader/index.js';
import _Footer from '../footer/index.js';
import _Empty from '../empty/index.js';
import _Link from '../link/index.js';
import { Form, FormItem } from '../form/index.js';
import { Icon } from '../icon/index.js';
import _BaseTable from '../table/index.js';
import { SideBar, SideBarItem } from '../side-bar/index.js';
import _Popover from '../popover/index.js';
import _Guide from '../guide/index.js';
import { ConfigProvider } from '../config-provider/index.js';

var components = /*#__PURE__*/Object.freeze({
	__proto__: null,
	Button: _Button,
	Divider: Divider,
	Calendar: _Calendar,
	Cell: _Cell,
	CellGroup: CellGroup,
	Checkbox: CheckBox,
	CheckboxGroup: CheckboxGroup,
	Input: _Input,
	Textarea: _Textarea,
	Overlay: Overlay,
	Progress: _Progress,
	Radio: _Radio,
	RadioGroup: RadioGroup,
	Rate: _Rate,
	Switch: _Switch,
	Stepper: Stepper,
	Slider: _Slider,
	Dialog: DialogPlugin,
	Popup: _Popup,
	ActionSheet: _ActionSheet,
	Badge: _Badge,
	Fab: _Fab,
	ImageViewer: _ImageViewer,
	Drawer: Drawer,
	Navbar: _Navbar,
	Message: MessagePlugin,
	DateTimePicker: _DateTimePicker,
	Toast: ToastPlugin,
	Search: _Search,
	Steps: Steps,
	StepItem: StepItem,
	Indexes: Indexes,
	IndexesAnchor: IndexesAnchor,
	Picker: Picker,
	Cascade: Cascade,
	Tag: Tag,
	CheckTag: CheckTag,
	TabBar: TabBar,
	TabBarItem: TabBarItem,
	DropdownMenu: DropdownMenu,
	DropdownItem: DropdownItem,
	Collapse: Collapse,
	CollapsePanel: CollapsePanel,
	Tabs: Tabs,
	TabPanel: TabPanel,
	Swiper: Swiper,
	SwiperItem: SwiperItem,
	CountDown: CountDown,
	Avatar: Avatar,
	AvatarGroup: AvatarGroup,
	Image: _Image,
	SwipeCell: SwipeCell,
	Skeleton: _Skeleton,
	Grid: Grid,
	GridItem: GridItem,
	Sticky: _Sticky,
	List: _List,
	Loading: _Loading,
	BackTop: _BackTop,
	PullDownRefresh: _PullDownRefresh,
	Upload: _Upload,
	NoticeBar: _NoticeBar,
	Result: _Result,
	Cascader: _Cascader,
	Footer: _Footer,
	Empty: _Empty,
	Link: _Link,
	Form: Form,
	FormItem: FormItem,
	Icon: Icon,
	Table: _BaseTable,
	SideBar: SideBar,
	SideBarItem: SideBarItem,
	TPopover: _Popover,
	Guide: _Guide,
	ConfigProvider: ConfigProvider
});

export { components as c };
