// tdesign-mobile-vue/lib/form/form-item-props.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var props = {
  for: {
    type: String,
    default: ""
  },
  help: {
    type: [String, Function]
  },
  label: {
    type: [String, Function],
    default: ""
  },
  labelAlign: {
    type: String,
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right", "top"].includes(val);
    }
  },
  labelWidth: {
    type: [String, Number]
  },
  contentAlign: {
    type: String,
    default: "left",
    validator: function validator(val) {
      if (!val) return true;
      return ["left", "right"].includes(val);
    }
  },
  name: {
    type: [String, Number]
  },
  requiredMark: {
    type: Boolean,
    default: void 0
  },
  rules: {
    type: Array
  },
  showErrorMessage: {
    type: Boolean,
    default: void 0
  },
  status: {
    type: String,
    default: ""
  },
  statusIcon: {
    type: [Boolean, Function],
    default: void 0
  },
  successBorder: Boolean,
  tips: {
    type: [String, Function]
  },
  arrow: Boolean
};

exports["default"] = props;
