// tdesign-mobile-vue/lib/form/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _asyncToGenerator = require('@babel/runtime/helpers/asyncToGenerator');
var vue = require('vue');
var _regeneratorRuntime = require('@babel/runtime/regenerator');
var isEmpty = require('lodash/isEmpty');
var isArray = require('lodash/isArray');
var isBoolean = require('lodash/isBoolean');
var isFunction = require('lodash/isFunction');
var form_props = require('./props.js');
var form_const = require('./const.js');
var config = require('../config.js');
var shared_dom = require('../shared/dom.js');
var hooks_tnode = require('../hooks/tnode.js');
var shared_render = require('../shared/render.js');
var _toConsumableArray = require('@babel/runtime/helpers/toConsumableArray');
var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var isNumber = require('lodash/isNumber');
var isString = require('lodash/isString');
var cloneDeep = require('lodash/cloneDeep');
var get = require('lodash/get');
var lodashSet = require('lodash/set');
var isNil = require('lodash/isNil');
var lodashTemplate = require('lodash/template');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var form_formModel = require('./form-model.js');
var form_formItemProps = require('./form-item-props.js');
var shared_renderTnode = require('../shared/render-tnode.js');
var shared_component = require('../shared/component.js');
require('../shared/util.js');
require('@babel/runtime/helpers/slicedToArray');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../hooks/render-tnode.js');
require('lodash/isObject');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2088457f.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _asyncToGenerator__default = /*#__PURE__*/_interopDefaultLegacy(_asyncToGenerator);
var _regeneratorRuntime__default = /*#__PURE__*/_interopDefaultLegacy(_regeneratorRuntime);
var isEmpty__default = /*#__PURE__*/_interopDefaultLegacy(isEmpty);
var isArray__default = /*#__PURE__*/_interopDefaultLegacy(isArray);
var isBoolean__default = /*#__PURE__*/_interopDefaultLegacy(isBoolean);
var isFunction__default = /*#__PURE__*/_interopDefaultLegacy(isFunction);
var _toConsumableArray__default = /*#__PURE__*/_interopDefaultLegacy(_toConsumableArray);
var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);
var isNumber__default = /*#__PURE__*/_interopDefaultLegacy(isNumber);
var isString__default = /*#__PURE__*/_interopDefaultLegacy(isString);
var cloneDeep__default = /*#__PURE__*/_interopDefaultLegacy(cloneDeep);
var get__default = /*#__PURE__*/_interopDefaultLegacy(get);
var lodashSet__default = /*#__PURE__*/_interopDefaultLegacy(lodashSet);
var isNil__default = /*#__PURE__*/_interopDefaultLegacy(isNil);
var lodashTemplate__default = /*#__PURE__*/_interopDefaultLegacy(lodashTemplate);

var prefix$1 = config["default"].prefix;
var name$1 = "".concat(prefix$1, "-form");
var requestSubmit = function requestSubmit(target) {
  if (!(target instanceof HTMLFormElement)) {
    throw new Error("target must be HTMLFormElement");
  }
  var submitter = document.createElement("input");
  submitter.type = "submit";
  submitter.hidden = true;
  target.appendChild(submitter);
  submitter.click();
  target.removeChild(submitter);
};
var _Form = vue.defineComponent({
  name: name$1,
  props: form_props["default"],
  setup: function setup(props2, _ref) {
    var expose = _ref.expose;
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _toRefs = vue.toRefs(props2),
      disabled = _toRefs.disabled,
      showErrorMessage = _toRefs.showErrorMessage,
      labelWidth = _toRefs.labelWidth,
      labelAlign = _toRefs.labelAlign,
      contentAlign = _toRefs.contentAlign,
      data = _toRefs.data,
      colon = _toRefs.colon,
      requiredMark = _toRefs.requiredMark,
      rules = _toRefs.rules,
      errorMessage = _toRefs.errorMessage,
      resetType = _toRefs.resetType;
    var formRef = vue.ref(null);
    var children = vue.ref([]);
    vue.provide("formDisabled", {
      disabled: disabled
    });
    vue.provide(form_const.FormInjectionKey, vue.reactive({
      showErrorMessage: showErrorMessage,
      labelWidth: labelWidth,
      labelAlign: labelAlign,
      contentAlign: contentAlign,
      data: data,
      colon: colon,
      requiredMark: requiredMark,
      rules: rules,
      errorMessage: errorMessage,
      resetType: resetType,
      children: children,
      renderContent: shared_render.renderContent
    }));
    var formClass = vue.computed(function () {
      return [name$1];
    });
    var needValidate = function needValidate(name2, fields) {
      if (!fields || !isArray__default["default"](fields)) return true;
      return fields.indexOf("".concat(name2)) !== -1;
    };
    var formatValidateResult = function formatValidateResult(validateResultList) {
      var result = validateResultList.reduce(function (r, err) {
        return Object.assign(r || {}, err);
      }, {});
      Object.keys(result).forEach(function (key) {
        if (result[key] === true) {
          delete result[key];
        }
      });
      return isEmpty__default["default"](result) ? true : result;
    };
    var validate = /*#__PURE__*/function () {
      var _ref2 = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee(param) {
        var _props2$onValidate;
        var _ref3, fields, _ref3$trigger, trigger, showErrorMessage2, list, arr, result;
        return _regeneratorRuntime__default["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _ref3 = param || {}, fields = _ref3.fields, _ref3$trigger = _ref3.trigger, trigger = _ref3$trigger === void 0 ? "all" : _ref3$trigger, showErrorMessage2 = _ref3.showErrorMessage;
              list = children.value.filter(function (child) {
                return isFunction__default["default"](child.validate) && needValidate(String(child.name), fields);
              }).map(function (child) {
                return child.validate(trigger, showErrorMessage2);
              });
              _context.next = 4;
              return Promise.all(list);
            case 4:
              arr = _context.sent;
              result = formatValidateResult(arr);
              (_props2$onValidate = props2.onValidate) === null || _props2$onValidate === void 0 || _props2$onValidate.call(props2, {
                validateResult: result
              });
              return _context.abrupt("return", result);
            case 8:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function validate(_x) {
        return _ref2.apply(this, arguments);
      };
    }();
    var getFirstError = function getFirstError(r) {
      var _r$Object$keys$, _Object$keys;
      if (isBoolean__default["default"](r)) return "";
      return (r === null || r === void 0 || (_r$Object$keys$ = r[(_Object$keys = Object.keys(r)) === null || _Object$keys === void 0 ? void 0 : _Object$keys[0]]) === null || _r$Object$keys$ === void 0 || (_r$Object$keys$ = _r$Object$keys$[0]) === null || _r$Object$keys$ === void 0 ? void 0 : _r$Object$keys$.message) || "";
    };
    var submitParams = vue.ref();
    var _onSubmit = function onSubmit(e) {
      if (props2.preventSubmitDefault && e) {
        shared_dom.preventDefault(e, true);
      }
      validate(submitParams.value).then(function (r) {
        var _props2$onSubmit;
        var firstError = getFirstError(r);
        (_props2$onSubmit = props2.onSubmit) === null || _props2$onSubmit === void 0 || _props2$onSubmit.call(props2, {
          validateResult: r,
          firstError: firstError
        });
      });
      submitParams.value = void 0;
    };
    var submit = /*#__PURE__*/function () {
      var _ref4 = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee2(params) {
        return _regeneratorRuntime__default["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              submitParams.value = params;
              requestSubmit(formRef.value);
            case 2:
            case "end":
              return _context2.stop();
          }
        }, _callee2);
      }));
      return function submit(_x2) {
        return _ref4.apply(this, arguments);
      };
    }();
    var resetParams = vue.ref();
    var _onReset = function onReset(e) {
      var _props2$onReset;
      if (props2.preventSubmitDefault && e) {
        shared_dom.preventDefault(e, true);
      }
      children.value.filter(function (child) {
        var _resetParams$value;
        return isFunction__default["default"](child.resetField) && needValidate(String(child.name), (_resetParams$value = resetParams.value) === null || _resetParams$value === void 0 ? void 0 : _resetParams$value.fields);
      }).forEach(function (child) {
        var _resetParams$value2;
        return child.resetField((_resetParams$value2 = resetParams.value) === null || _resetParams$value2 === void 0 ? void 0 : _resetParams$value2.type);
      });
      resetParams.value = void 0;
      (_props2$onReset = props2.onReset) === null || _props2$onReset === void 0 || _props2$onReset.call(props2, {
        e: e
      });
    };
    var reset = function reset(params) {
      resetParams.value = params;
      formRef.value.reset();
    };
    var clearValidate = function clearValidate(fields) {
      children.value.forEach(function (child) {
        if (isFunction__default["default"](child.resetHandler) && needValidate(String(child.name), fields)) {
          child.resetHandler();
        }
      });
    };
    var setValidateMessage = function setValidateMessage(validateMessage) {
      var keys = Object.keys(validateMessage);
      if (!keys.length) return;
      var list = children.value.filter(function (child) {
        return isFunction__default["default"](child.setValidateMessage) && keys.includes("".concat(child.name));
      }).map(function (child) {
        return child.setValidateMessage(validateMessage["".concat(child.name)]);
      });
      Promise.all(list);
    };
    expose({
      validate: validate,
      submit: submit,
      reset: reset,
      clearValidate: clearValidate,
      setValidateMessage: setValidateMessage
    });
    return function () {
      return vue.createVNode("form", {
        "ref": formRef,
        "class": formClass.value,
        "onSubmit": function onSubmit(e) {
          return _onSubmit(e);
        },
        "onReset": function onReset(e) {
          return _onReset(e);
        }
      }, [renderTNodeJSX("default")]);
    };
  }
});

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty__default["default"](e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix = config["default"].prefix;
var name = "".concat(prefix, "-form-item");
var classPrefix = "".concat(prefix, "-form__item");
var _FormItem = vue.defineComponent({
  name: name,
  components: {
    TNode: shared_renderTnode["default"]
  },
  props: form_formItemProps["default"],
  setup: function setup(props2, _ref) {
    var slots = _ref.slots;
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var _toRefs = vue.toRefs(props2),
      name2 = _toRefs.name;
    var form = vue.inject(form_const.FormInjectionKey, void 0);
    var extraNode = vue.computed(function () {
      var _list$;
      var list = errorList.value;
      if (showErrorMessage.value && (_list$ = list[0]) !== null && _list$ !== void 0 && _list$.message) {
        var _list$2;
        return (_list$2 = list[0]) === null || _list$2 === void 0 ? void 0 : _list$2.message;
      }
      if (successList.value.length) {
        return successList.value[0].message;
      }
      return null;
    });
    var formItemClass = vue.computed(function () {
      return ["".concat(prefix, "-form__item"), "".concat(prefix, "-form__item--bordered"), "".concat(prefix, "-form--").concat(labelAlign.value), "".concat(prefix, "-form-item__").concat(props2.name)];
    });
    var needRequiredMark = vue.computed(function () {
      var _props2$requiredMark;
      var requiredMark = (_props2$requiredMark = props2.requiredMark) !== null && _props2$requiredMark !== void 0 ? _props2$requiredMark : form === null || form === void 0 ? void 0 : form.requiredMark;
      var isRequired = innerRules.value.filter(function (rule) {
        return rule.required;
      }).length > 0;
      return requiredMark !== null && requiredMark !== void 0 ? requiredMark : isRequired;
    });
    var hasLabel = vue.computed(function () {
      return slots.label || props2.label;
    });
    var hasColon = vue.computed(function () {
      return !!(form !== null && form !== void 0 && form.colon && hasLabel.value);
    });
    var FROM_LABEL = "".concat(prefix, "-form__label");
    var labelAlign = vue.computed(function () {
      return isNil__default["default"](props2.labelAlign) ? form === null || form === void 0 ? void 0 : form.labelAlign : props2.labelAlign;
    });
    var labelWidth = vue.computed(function () {
      return isNil__default["default"](props2.labelWidth) ? form === null || form === void 0 ? void 0 : form.labelWidth : props2.labelWidth;
    });
    var contentAlign = vue.computed(function () {
      return isNil__default["default"](props2.contentAlign) ? form === null || form === void 0 ? void 0 : form.contentAlign : props2.contentAlign;
    });
    var labelClasses = vue.computed(function () {
      return ["".concat(prefix, "-form__label"), _defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"](_defineProperty__default["default"]({}, "".concat(FROM_LABEL, "--required"), needRequiredMark.value), "".concat(FROM_LABEL, "--colon"), hasColon.value), "".concat(FROM_LABEL, "--top"), hasLabel.value && (labelAlign.value === "top" || !labelWidth.value)), "".concat(FROM_LABEL, "--left"), labelAlign.value === "left" && labelWidth.value), "".concat(FROM_LABEL, "--right"), labelAlign.value === "right" && labelWidth.value)];
    });
    var labelStyle = vue.computed(function () {
      if (labelWidth.value && labelAlign.value !== "top") {
        return isNumber__default["default"](labelWidth.value) ? {
          width: "".concat(labelWidth.value, "px")
        } : {
          width: labelWidth.value
        };
      }
      return {};
    });
    var freeShowErrorMessage = vue.ref(false);
    var showErrorMessage = vue.computed(function () {
      if (isBoolean__default["default"](freeShowErrorMessage.value)) return freeShowErrorMessage.value;
      if (isBoolean__default["default"](props2.showErrorMessage)) return props2.showErrorMessage;
      return form === null || form === void 0 ? void 0 : form.showErrorMessage;
    });
    var errorClasses = vue.computed(function () {
      if (!showErrorMessage.value) return "";
      if (!errorList.value.length) return "";
      var type = errorList.value[0].type || "error";
      return type === "error" ? "".concat(classPrefix, "--error") : "".concat(classPrefix, "--warning");
    });
    var contentClasses = vue.computed(function () {
      return ["".concat(prefix, "-form__controls"), errorClasses.value];
    });
    var contentSlotClasses = vue.computed(function () {
      return ["".concat(prefix, "-form__controls-content"), "".concat(prefix, "-form__controls--").concat(contentAlign.value)];
    });
    var contentStyle = vue.computed(function () {
      var contentStyle2 = {};
      if (labelWidth.value && labelAlign.value !== "top") {
        if (isNumber__default["default"](labelWidth.value)) {
          contentStyle2 = {
            marginLeft: "".concat(labelWidth.value, "px")
          };
        } else {
          contentStyle2 = {
            marginLeft: labelWidth.value
          };
        }
      }
      return contentStyle2;
    });
    var errorList = vue.ref([]);
    var successList = vue.ref([]);
    var verifyStatus = vue.ref(form_const.ValidateStatus.TO_BE_VALIDATED);
    var resetValidating = vue.ref(false);
    var needResetField = vue.ref(false);
    var resetHandler = function resetHandler() {
      needResetField.value = false;
      errorList.value = [];
      successList.value = [];
      verifyStatus.value = form_const.ValidateStatus.TO_BE_VALIDATED;
    };
    var getEmptyValue = function getEmptyValue() {
      var type = Object.prototype.toString.call(get__default["default"](form === null || form === void 0 ? void 0 : form.data, "".concat(props2.name)));
      var emptyValue;
      if (type === "[object String]") {
        emptyValue = "";
      }
      if (type === "[object Array]") {
        emptyValue = [];
      }
      if (type === "[object Object]") {
        emptyValue = {};
      }
      return emptyValue;
    };
    var resetField = /*#__PURE__*/function () {
      var _ref3 = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee() {
        var resetType,
          _args = arguments;
        return _regeneratorRuntime__default["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              resetType = _args.length > 0 && _args[0] !== undefined ? _args[0] : form === null || form === void 0 ? void 0 : form.resetType;
              if (props2.name) {
                _context.next = 3;
                break;
              }
              return _context.abrupt("return", null);
            case 3:
              if (resetType === "empty") {
                lodashSet__default["default"](form === null || form === void 0 ? void 0 : form.data, props2.name, getEmptyValue());
              } else if (resetType === "initial") {
                lodashSet__default["default"](form === null || form === void 0 ? void 0 : form.data, props2.name, initialValue.value);
              }
              _context.next = 6;
              return vue.nextTick();
            case 6:
              if (resetValidating.value) {
                needResetField.value = true;
              } else {
                resetHandler();
              }
            case 7:
            case "end":
              return _context.stop();
          }
        }, _callee);
      }));
      return function resetField() {
        return _ref3.apply(this, arguments);
      };
    }();
    var errorMessages = vue.computed(function () {
      return (form === null || form === void 0 ? void 0 : form.errorMessage) || {};
    });
    var innerRules = vue.computed(function () {
      var _props2$rules;
      if ((_props2$rules = props2.rules) !== null && _props2$rules !== void 0 && _props2$rules.length) return props2.rules;
      if (!props2.name) return [];
      var index = "".concat(props2.name).lastIndexOf(".") || -1;
      var pRuleName = "".concat(props2.name).slice(index + 1);
      return get__default["default"](form === null || form === void 0 ? void 0 : form.rules, props2.name) || get__default["default"](form === null || form === void 0 ? void 0 : form.rules, pRuleName) || [];
    });
    var analysisValidateResult = /*#__PURE__*/function () {
      var _ref4 = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee2(trigger) {
        var _result$rules;
        var result;
        return _regeneratorRuntime__default["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              result = {
                successList: [],
                errorList: [],
                rules: [],
                resultList: [],
                allowSetValue: false
              };
              result.rules = trigger === "all" ? innerRules.value : innerRules.value.filter(function (item) {
                return (item.trigger || "change") === trigger;
              });
              if (!(innerRules.value.length && !((_result$rules = result.rules) !== null && _result$rules !== void 0 && _result$rules.length))) {
                _context2.next = 4;
                break;
              }
              return _context2.abrupt("return", result);
            case 4:
              result.allowSetValue = true;
              _context2.next = 7;
              return form_formModel.validate(value.value, result.rules);
            case 7:
              result.resultList = _context2.sent;
              result.errorList = result.resultList.filter(function (item) {
                return item.result !== true;
              }).map(function (item) {
                Object.keys(item).forEach(function (key) {
                  if (!item.message && errorMessages.value[key]) {
                    var compiled = lodashTemplate__default["default"](errorMessages.value[key]);
                    var name3 = isString__default["default"](props2.label) ? props2.label : props2.name;
                    item.message = compiled({
                      name: name3,
                      validate: item[key]
                    });
                  }
                });
                return item;
              });
              result.successList = result.resultList.filter(function (item) {
                return item.result === true && item.message && item.type === "success";
              });
              return _context2.abrupt("return", result);
            case 11:
            case "end":
              return _context2.stop();
          }
        }, _callee2);
      }));
      return function analysisValidateResult(_x) {
        return _ref4.apply(this, arguments);
      };
    }();
    var validateHandler = /*#__PURE__*/function () {
      var _ref5 = _asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee3(trigger, showErrorMessage2) {
        var _yield$analysisValida, innerSuccessList, innerErrorList, rules, resultList, allowSetValue;
        return _regeneratorRuntime__default["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              resetValidating.value = true;
              freeShowErrorMessage.value = showErrorMessage2;
              _context3.next = 4;
              return analysisValidateResult(trigger);
            case 4:
              _yield$analysisValida = _context3.sent;
              innerSuccessList = _yield$analysisValida.successList;
              innerErrorList = _yield$analysisValida.errorList;
              rules = _yield$analysisValida.rules;
              resultList = _yield$analysisValida.resultList;
              allowSetValue = _yield$analysisValida.allowSetValue;
              if (allowSetValue) {
                successList.value = innerSuccessList || [];
                errorList.value = innerErrorList || [];
              }
              if (rules.length) {
                verifyStatus.value = innerErrorList !== null && innerErrorList !== void 0 && innerErrorList.length ? form_const.ValidateStatus.FAIL : form_const.ValidateStatus.SUCCESS;
              }
              if (needResetField.value) {
                resetHandler();
              }
              resetValidating.value = false;
              return _context3.abrupt("return", _defineProperty__default["default"]({}, "".concat(name2.value), (innerErrorList === null || innerErrorList === void 0 ? void 0 : innerErrorList.length) === 0 ? true : resultList));
            case 15:
            case "end":
              return _context3.stop();
          }
        }, _callee3);
      }));
      return function validateHandler(_x2, _x3) {
        return _ref5.apply(this, arguments);
      };
    }();
    var setValidateMessage = function setValidateMessage(validateMessage) {
      if (!validateMessage && !isArray__default["default"](validateMessage)) return;
      if (validateMessage.length === 0) {
        errorList.value = [];
        verifyStatus.value = form_const.ValidateStatus.SUCCESS;
      }
      errorList.value = validateMessage.map(function (item) {
        return _objectSpread(_objectSpread({}, item), {}, {
          result: false
        });
      });
      verifyStatus.value = form_const.ValidateStatus.FAIL;
    };
    var value = vue.computed(function () {
      return (form === null || form === void 0 ? void 0 : form.data) && get__default["default"](form === null || form === void 0 ? void 0 : form.data, "".concat(name2.value));
    });
    var initialValue = vue.ref(void 0);
    var context = vue.reactive({
      name: name2,
      resetHandler: resetHandler,
      resetField: resetField,
      validate: validateHandler,
      setValidateMessage: setValidateMessage
    });
    vue.onMounted(function () {
      initialValue.value = cloneDeep__default["default"](value.value);
      form === null || form === void 0 || form.children.push(context);
    });
    vue.onBeforeUnmount(function () {
      if (form) form.children = form === null || form === void 0 ? void 0 : form.children.filter(function (ctx) {
        return ctx !== context;
      });
    });
    vue.watch(value, /*#__PURE__*/_asyncToGenerator__default["default"]( /*#__PURE__*/_regeneratorRuntime__default["default"].mark(function _callee4() {
      return _regeneratorRuntime__default["default"].wrap(function _callee4$(_context4) {
        while (1) switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return validateHandler("change");
          case 2:
          case "end":
            return _context4.stop();
        }
      }, _callee4);
    })), {
      deep: true
    });
    vue.watch(function () {
      return [props2.name, JSON.stringify(props2.rules)].join(",");
    }, function () {
      validateHandler("change");
    });
    return function () {
      var renderRightIconContent = function renderRightIconContent() {
        if (!props2.arrow) {
          return null;
        }
        return vue.h(tdesignIconsVueNext.ChevronRightIcon, {
          size: "24px",
          color: "rgba(0, 0, 0, .4)"
        });
      };
      var renderLabelContent = function renderLabelContent() {
        if (Number(labelWidth.value) === 0) {
          return null;
        }
        return renderTNodeJSX("label");
      };
      var renderHelpNode = function renderHelpNode() {
        var helpNode = renderTNodeJSX("help");
        if (!helpNode) {
          return null;
        }
        return vue.createVNode("div", {
          "class": ["".concat(classPrefix, "-help"), "".concat(prefix, "-form__controls--").concat(contentAlign.value)]
        }, [helpNode]);
      };
      var renderExtraNode = function renderExtraNode() {
        if (!extraNode.value) {
          return null;
        }
        return vue.createVNode("div", {
          "class": ["".concat(classPrefix, "-extra"), "".concat(prefix, "-form__controls--").concat(contentAlign.value)]
        }, [extraNode.value]);
      };
      return vue.createVNode("div", {
        "class": [].concat(_toConsumableArray__default["default"](formItemClass.value), [renderHelpNode() ? "".concat(prefix, "-form__item-with-help") : ""])
      }, [vue.createVNode("div", {
        "class": ["".concat(classPrefix, "-wrap"), "".concat(classPrefix, "--").concat(labelAlign.value)]
      }, [vue.createVNode("div", {
        "class": labelClasses.value,
        "style": labelStyle.value
      }, [vue.createVNode("label", {
        "for": props2.for
      }, [renderLabelContent()])]), vue.createVNode("div", {
        "class": contentClasses.value,
        "style": contentStyle.value
      }, [vue.createVNode("div", {
        "class": contentSlotClasses.value
      }, [renderTNodeJSX("default")]), renderHelpNode(), renderExtraNode()])]), renderRightIconContent()]);
    };
  }
});

var Form = shared_component.withInstall(_Form);
var FormItem = shared_component.withInstall(_FormItem);

exports.Form = Form;
exports.FormItem = FormItem;
