// tdesign-mobile-vue/esm/cascader/props.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

var TdCascaderProps = {
  closeBtn: {
    type: [Boolean, Function],
    default: true
  },
  keys: {
    type: Object
  },
  options: {
    type: Array,
    default: function _default() {
      return [];
    }
  },
  subTitles: {
    type: Array,
    default: function _default() {
      return [];
    }
  },
  theme: {
    type: String,
    default: "step",
    validator: function validator(val) {
      if (!val) return true;
      return ["step", "tab"].includes(val);
    }
  },
  title: {
    type: [String, Function]
  },
  value: {
    type: [String, Number],
    default: void 0
  },
  modelValue: {
    type: [String, Number],
    default: void 0
  },
  defaultValue: {
    type: [String, Number]
  },
  placeholder: {
    type: [String, Function]
  },
  visible: Boolean,
  checkStrictly: {
    type: Boolean,
    default: false
  },
  onChange: Function,
  onClose: Function,
  onPick: Function
};

export { TdCascaderProps as default };
