// tdesign-mobile-vue/es/tabs/index.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

import _toConsumableArray from '@babel/runtime/helpers/toConsumableArray';
import _slicedToArray from '@babel/runtime/helpers/slicedToArray';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { defineComponent, createVNode, computed, ref, toRefs, Fragment, onMounted, onBeforeUnmount, onActivated, watch, nextTick, h, provide, readonly, mergeProps, inject, withDirectives, vShow } from 'vue';
import isFunction from 'lodash/isFunction';
import config from '../config.js';
import props$1 from './props.js';
import props from './tab-panel-props.js';
import { useContent, useTNodeJSX } from '../hooks/tnode.js';
import '../shared/index.js';
import { preventDefault } from '../shared/dom.js';
import ClASSNAMES from '../shared/constants.js';
import _Sticky from '../sticky/index.js';
import _Badge from '../badge/index.js';
import { usePrefixClass } from '../hooks/useClass.js';
import { useVModel } from '../shared/useVModel/index.js';
import './style/css.js';
import './type.js';
import { withInstall } from '../shared/component.js';
import 'lodash/camelCase';
import 'lodash/kebabCase';
import '../hooks/render-tnode.js';
import 'lodash/isEmpty';
import 'lodash/isString';
import 'lodash/isObject';
import '../shared/functions.js';
import '../shared/util.js';
import 'lodash/isNumber';
import '../shared/render.js';
import '@babel/runtime/helpers/typeof';
import '../shared/render-tnode.js';
import '../shared/useToggle/index.js';
import '../shared/useCountDown/index.js';
import '@babel/runtime/helpers/asyncToGenerator';
import '@babel/runtime/regenerator';
import '@vueuse/core';
import '../shared/useCountDown/utils.js';
import '../shared/useDefault/index.js';
import '../shared/useChildSlots/index.js';
import '../shared/useTouch/index.js';
import '../shared/useScrollParent/index.js';
import '../shared/useExpose/index.js';
import '../shared/useTest/index.js';
import '../shared/useClickAway/index.js';
import 'lodash/isArray';
import '../shared/useGesture/index.js';
import '@use-gesture/vanilla';
import '../shared/hover.js';
import '../sticky/props.js';
import '../sticky/type.js';
import '../badge/props.js';
import '../badge/type.js';
import '../config-provider/useConfig.js';
import 'lodash/cloneDeep';
import '../config-provider/context.js';
import 'lodash/mergeWith';
import 'lodash/merge';
import '../_common/js/global-config/mobile/default-config.js';
import '../_common/js/global-config/mobile/locale/zh_CN.js';
import '../_chunks/dep-4c6045f6.js';
import '../_chunks/dep-1d280204.js';
import 'dayjs';
import '../_chunks/dep-5101c61a.js';
import '../config-provider/type.js';

var prefix$2 = config.prefix;
var TTabNavItem = defineComponent({
  name: "".concat(prefix$2, "-tab-nav"),
  props: {
    label: props.label
  },
  setup: function setup(props) {
    var renderTNodeContent = useContent();
    return function () {
      return createVNode("div", null, [renderTNodeContent("default", "label")]);
    };
  }
});

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var prefix$1 = config.prefix;
var name$1 = "".concat(prefix$1, "-tabs");
var LocalTabs = defineComponent({
  name: name$1,
  props: props$1,
  setup: function setup(props2) {
    var renderTNodeJSX = useTNodeJSX();
    var tabsClass = usePrefixClass("tabs");
    var stickyProps = computed(function () {
      return _objectSpread(_objectSpread({}, props2.stickyProps), {}, {
        disabled: !props2.sticky
      });
    });
    var activeClass = "".concat(tabsClass.value, "__item--active");
    var disabledClass = "".concat(tabsClass.value, "__item--disabled");
    var tabsClasses = computed(function () {
      return ["".concat(tabsClass.value), props2.size && ClASSNAMES.SIZE[props2.size]];
    });
    var navClasses = ref(["".concat(tabsClass.value, "__nav")]);
    var startX = ref(0);
    var startY = ref(0);
    var endX = ref(0);
    var endY = ref(0);
    var canMove = ref(true);
    var tabIndex = computed(function () {
      var index = 0;
      for (var i = 0; i < itemProps.value.length; i++) {
        if (itemProps.value[i].value === currentValue.value) {
          index = i;
          break;
        }
      }
      return index;
    });
    var _toRefs = toRefs(props2),
      value = _toRefs.value,
      modelValue = _toRefs.modelValue;
    var _useVModel = useVModel(value, modelValue, props2.defaultValue, props2.onChange),
      _useVModel2 = _slicedToArray(_useVModel, 2),
      currentValue = _useVModel2[0],
      setCurrentValue = _useVModel2[1];
    var itemProps = computed(function () {
      var _children;
      if (props2.list) {
        return props2.list;
      }
      var children = renderTNodeJSX("default");
      var res = [];
      var _label = [];
      (_children = children) === null || _children === void 0 || _children.forEach(function (child) {
        var _child$children;
        if (child.type === Fragment) {
          res.push.apply(res, _toConsumableArray(child.children));
        } else {
          res.push(child);
        }
        if ((_child$children = child.children) !== null && _child$children !== void 0 && _child$children.label) {
          _label.push(child.children.label()[0] || null);
        }
      });
      children = res.filter(function (child) {
        return child.type.name === "".concat(prefix$1, "-tab-panel");
      });
      return children.map(function (item, index) {
        return _objectSpread(_objectSpread({}, item.props), {}, {
          label: function label() {
            return _label[index] || item.props.label;
          }
        });
      });
    });
    var valueList = computed(function () {
      return itemProps.value.map(function (v) {
        return v.value;
      });
    });
    var currentIndex = computed(function () {
      return valueList.value.indexOf(currentValue.value);
    });
    var navScroll = ref();
    var navWrap = ref();
    var navLine = ref();
    var lineStyle = ref();
    var moveToActiveTab = function moveToActiveTab() {
      if (navWrap.value && navLine.value && props2.showBottomLine) {
        var tab = navWrap.value.querySelector(".".concat(activeClass));
        if (!tab) return;
        var line = navLine.value;
        var tabInner = tab.querySelector(".".concat(prefix$1, "-badge"));
        var style = {};
        if (props2.bottomLineMode === "auto") {
          style.width = "".concat(Number(tabInner === null || tabInner === void 0 ? void 0 : tabInner.offsetWidth), "px");
          style.transform = "translateX(".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetLeft) + Number(tabInner === null || tabInner === void 0 ? void 0 : tabInner.offsetLeft), "px)");
        } else if (props2.bottomLineMode === "full") {
          style.width = "".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetWidth), "px");
          style.transform = "translateX(".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetLeft), "px)");
        } else {
          style.transform = "translateX(".concat(Number(tab === null || tab === void 0 ? void 0 : tab.offsetLeft) + (Number(tab === null || tab === void 0 ? void 0 : tab.offsetWidth) - Number(line === null || line === void 0 ? void 0 : line.offsetWidth)) / 2, "px)");
        }
        if (props2.animation) {
          style.transitionDuration = "".concat(props2.animation.duration, "ms");
        }
        lineStyle.value = style;
      }
    };
    onMounted(function () {
      window.addEventListener("resize", moveToActiveTab, false);
      setTimeout(function () {
        moveToActiveTab();
      }, 300);
    });
    onBeforeUnmount(function () {
      window.removeEventListener("resize", moveToActiveTab);
    });
    onActivated(function () {
      moveToActiveTab();
    });
    watch(value, function () {
      nextTick(function () {
        moveToActiveTab();
      });
    });
    var handleTabClick = function handleTabClick(event, item) {
      var _props2$onClick;
      var value2 = item.value,
        disabled = item.disabled;
      if (disabled || currentValue.value === value2) {
        return false;
      }
      var label = isFunction(item.label) ? item.label(h).toString() : item.label;
      setCurrentValue(value2, label);
      (_props2$onClick = props2.onClick) === null || _props2$onClick === void 0 || _props2$onClick.call(props2, value2, label);
      nextTick(function () {
        moveToActiveTab();
      });
    };
    var handlerScroll = function handlerScroll(context) {
      var scrollTop = context.scrollTop,
        isFixed = context.isFixed;
      if (props2.stickyProps) {
        var _props2$onScroll;
        (_props2$onScroll = props2.onScroll) === null || _props2$onScroll === void 0 || _props2$onScroll.call(props2, scrollTop, isFixed);
      }
    };
    var handleTouchstart = function handleTouchstart(e) {
      if (!props2.swipeable) return;
      startX.value = e.targetTouches[0].clientX;
      startY.value = e.targetTouches[0].clientY;
    };
    var handleTouchmove = function handleTouchmove(e) {
      if (!props2.swipeable) return;
      if (!canMove.value) return;
      endX.value = e.targetTouches[0].clientX;
      endY.value = e.targetTouches[0].clientY;
      var dValueX = Math.abs(startX.value - endX.value);
      var dValueY = Math.abs(startY.value - endY.value);
      if (tabIndex.value >= 0 && tabIndex.value < itemProps.value.length) {
        if (dValueX > dValueY) {
          preventDefault(e, false);
          if (dValueX <= 40) return;
          if (startX.value > endX.value) {
            if (tabIndex.value >= itemProps.value.length - 1) return;
            canMove.value = false;
            handleTabClick(e, itemProps.value[tabIndex.value + 1]);
          } else if (startX.value < endX.value) {
            if (tabIndex.value <= 0) return;
            canMove.value = false;
            handleTabClick(e, itemProps.value[tabIndex.value - 1]);
          }
        }
      }
    };
    var handleTouchend = function handleTouchend() {
      if (!props2.swipeable) return;
      canMove.value = true;
      startX.value = 0;
      endX.value = 0;
      startY.value = 0;
      endY.value = 0;
    };
    provide("currentValue", readonly(currentValue));
    var readerNav = function readerNav() {
      return itemProps.value.map(function (item, index) {
        var badgeProps = item["badge-props"] || item.badgeProps;
        return createVNode("div", {
          "class": _defineProperty(_defineProperty(_defineProperty(_defineProperty(_defineProperty({}, "".concat(tabsClass.value, "__item ").concat(tabsClass.value, "__item--top"), true), "".concat(tabsClass.value, "__item--evenly"), props2.spaceEvenly), activeClass, item.value === currentValue.value), disabledClass, item.disabled), "".concat(tabsClass.value, "__item--").concat(props2.theme), true),
          "onClick": function onClick(e) {
            return handleTabClick(e, item);
          }
        }, [createVNode(_Badge, badgeProps, {
          default: function _default() {
            return [createVNode("div", {
              "class": _defineProperty(_defineProperty({}, "".concat(tabsClass.value, "__item-inner ").concat(tabsClass.value, "__item-inner--").concat(props2.theme), true), "".concat(tabsClass.value, "__item-inner--active"), props2.theme === "tag" && item.value === currentValue.value)
            }, [createVNode(TTabNavItem, {
              "label": item.label
            }, null)])];
          }
        }), props2.theme === "card" && index === currentIndex.value - 1 && createVNode("div", {
          "class": "".concat(tabsClass.value, "__item-prefix")
        }, null), props2.theme === "card" && index === currentIndex.value + 1 && createVNode("div", {
          "class": "".concat(tabsClass.value, "__item-suffix")
        }, null)]);
      });
    };
    return function () {
      return createVNode("div", {
        "class": tabsClasses.value
      }, [createVNode(_Sticky, mergeProps(stickyProps.value, {
        "onScroll": handlerScroll
      }), {
        default: function _default() {
          return [createVNode("div", {
            "class": navClasses.value
          }, [createVNode("div", {
            "ref": navScroll,
            "class": "".concat(tabsClass.value, "__scroll ").concat(tabsClass.value, "__scroll--top ").concat(tabsClass.value, "__scroll--").concat(props2.theme)
          }, [createVNode("div", {
            "ref": navWrap,
            "class": "".concat(tabsClass.value, "__wrapper ").concat(tabsClass.value, "__wrapper--").concat(props2.theme)
          }, [readerNav(), props2.theme === "line" && props2.showBottomLine && createVNode("div", {
            "ref": navLine,
            "class": "".concat(tabsClass.value, "__track ").concat(tabsClass.value, "__track--top"),
            "style": lineStyle.value
          }, null)])])])];
        }
      }), createVNode("div", {
        "class": "".concat(tabsClass.value, "__content"),
        "onTouchstart": handleTouchstart,
        "onTouchmove": handleTouchmove,
        "onTouchend": handleTouchend
      }, [renderTNodeJSX("default")])]);
    };
  }
});

var prefix = config.prefix;
var name = "".concat(prefix, "-tab-panel");
var LocalTabPanel = defineComponent({
  name: name,
  props: props,
  setup: function setup(props2) {
    var renderTNodeContent = useContent();
    var tabPanelClass = usePrefixClass("tab-panel");
    var tabsClass = usePrefixClass("tabs");
    var currentValue = inject("currentValue");
    var isActive = computed(function () {
      return currentValue.value === props2.value;
    });
    var tabPanelClasses = computed(function () {
      return ["".concat(tabPanelClass.value), "".concat(tabsClass.value, "__panel")];
    });
    var isMount = ref(props2.lazy ? isActive.value : true);
    watch(isActive, function () {
      if (isActive.value) {
        if (!isMount.value) {
          isMount.value = true;
        }
      } else if (props2.destroyOnHide) {
        isMount.value = false;
      }
    }, {
      immediate: true
    });
    return function () {
      if (!isMount.value) return null;
      return withDirectives(createVNode("div", {
        "class": tabPanelClasses.value
      }, [renderTNodeContent("default", "panel")]), [[vShow, isActive.value]]);
    };
  }
});

var Tabs = withInstall(LocalTabs);
var TabPanel = withInstall(LocalTabPanel);

export { TabPanel, Tabs };
