// tdesign-mobile-vue/lib/config-provider/config-provider.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var config = require('../config.js');
var configProvider_useConfig = require('./useConfig.js');
require('lodash/isFunction');
require('lodash/cloneDeep');
require('lodash/isString');
require('./context.js');
require('lodash/mergeWith');
require('lodash/merge');
require('lodash/isArray');
require('../_common/js/global-config/mobile/default-config.js');
require('../_common/js/global-config/mobile/locale/zh_CN.js');
require('../_chunks/dep-191173f8.js');
require('@babel/runtime/helpers/typeof');
require('../_chunks/dep-2088457f.js');
require('dayjs');

var prefix = config["default"].prefix;
var name = "".concat(prefix, "-config-provider");
var configProviderProps = {
  globalConfig: Object
};
var script = vue.defineComponent({
  name: name,
  props: configProviderProps,
  setup: function setup(props) {
    configProvider_useConfig.provideConfig(props);
    return {
      name: name
    };
  }
});

function render(_ctx, _cache, $props, $setup, $data, $options) {
  return vue.renderSlot(_ctx.$slots, "default");
}

script.render = render;

exports.configProviderProps = configProviderProps;
exports["default"] = script;
