// tdesign-mobile-vue/lib/button/props.js
/**
 * tdesign v1.2.3
 * (c) 2024 TDesign Group
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var ButtonProps = {
  block: Boolean,
  content: {
    type: [String, Function]
  },
  disabled: Boolean,
  ghost: Boolean,
  icon: {
    type: Function
  },
  loading: Boolean,
  loadingProps: {
    type: Object
  },
  shape: {
    type: String,
    default: "rectangle",
    validator: function validator(val) {
      if (!val) return true;
      return ["rectangle", "square", "round", "circle"].includes(val);
    }
  },
  size: {
    type: String,
    default: "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["extra-small", "small", "medium", "large"].includes(val);
    }
  },
  suffix: {
    type: Function
  },
  theme: {
    type: String,
    default: "default",
    validator: function validator(val) {
      if (!val) return true;
      return ["default", "primary", "danger", "light"].includes(val);
    }
  },
  type: {
    type: String,
    default: "button",
    validator: function validator(val) {
      if (!val) return true;
      return ["submit", "reset", "button"].includes(val);
    }
  },
  variant: {
    type: String,
    default: "base",
    validator: function validator(val) {
      if (!val) return true;
      return ["base", "outline", "text"].includes(val);
    }
  },
  onClick: Function
};

exports["default"] = ButtonProps;
